### PR TITLE
remove FindUHDM.cmake, no longer necessary

### PR DESCRIPTION
**Pull request by timkpaine**

### CODE REVIEW

**Comment by hzeller** [comment]
When I install UHDM, it generates only one `*.cmake` file, a `UHDMConfig.cmake` in the install location, which mostly attempts to include UHDM.cmake, which doesn't exist. So it looks like it won't know anything about version, include and library locations, and anything else relevant.
 
It feels like something is missing. Are you working on some changes on UHDM alongside ?

```
$ cd UHDM

$ PREFIX=/tmp/uhdm-test make install

$ find /tmp/uhdm-test -name "*.cmake"
/tmp/uhdm-test/lib/cmake/UHDMConfig.cmake

$ cat /tmp/uhdm-test/lib/cmake/UHDMConfig.cmake

####### Expanded from @PACKAGE_INIT@ by configure_package_config_file() #######
####### Any changes to this file will be overwritten by the next CMake run ####
####### The input file was Config.cmake.in                            ########

get_filename_component(PACKAGE_PREFIX_DIR "${CMAKE_CURRENT_LIST_DIR}/../" ABSOLUTE)

####################################################################################

include ( "${CMAKE_CURRENT_LIST_DIR}/UHDM.cmake" )
```

**Comment by hzeller** [comment]
An idea to make the vendored testing more bullet-proof: To make sure that we don't pick up anything accidentally from the CI machine (which are not always very pristine and might contain gtest or flatbuffers etc.), maybe we should have a `DEPENDENCY_INSTALL_DIR` in the `Install vendored dependencies` section of the CI which is then then `PREFIX` for all the vendored binaries - and then use that to assemble search paths for cmake in the actual Surelog build (and in the UHDM build to find the Capnp cmake location).

That way we know that we have a robust installation that don't accidentally depends on things the CI machine might have, or some hidden dependencies on `/usr` or `/usr/local`.

**Comment by hzeller** [comment]
Are you working on a fix for the incomplete cmake file in UHDM, @timkpaine ?

**Comment by timkpaine** [comment]
Almost done, will do PR to UHDM first but in testing against Surelog locally so that can come right after

**Comment by timkpaine** [comment]
needs https://github.com/chipsalliance/UHDM/pull/903

**Comment by hzeller** [comment]
Alright #3564 is merged, so if you rebase to that, you'll get the latest UHDM that contains your UHDM updates.

There is this `test_install` target in the toplevel makefile, which should work in the end. I suspect the makefile it needs some setting of the `CMAKE_MODULE_PATH` that is based on the `$(PREFIX)`
```
PREFIX=/tmp/some_directory make install
# compile and install...

PREFIX=/tmp/some_directory make test_install
cmake -DCMAKE_BUILD_TYPE=Release -DINSTALL_DIR=/tmp/s1 -S tests/TestInstall -B tests/TestInstall/build
CMake Error at CMakeLists.txt:27 (find_package):
  By not providing "FindSurelog.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Surelog", but
  CMake did not find one.

  Could not find a package configuration file provided by "Surelog" with any
  of the following names:

    SurelogConfig.cmake
    surelog-config.cmake

  Add the installation prefix of "Surelog" to CMAKE_PREFIX_PATH or set
  "Surelog_DIR" to a directory containing one of the above files.  If
  "Surelog" provides a separate development package or SDK, be sure it has
  been installed.
```

**Comment by hzeller** [comment]
Once this is working and submitted, I'll tag UHDM and Surelog to 1.52

**Comment by timkpaine** [comment]
Still working through this, didn't have as much time over the weekend as I would've liked

**Comment by timkpaine** [comment]
The cmake changes work, I just want to make sure the tests are really testing everything

**Comment by timkpaine** [comment]
home stretch

File: cmake/configs/SurelogConfig.cmake.in [hunk 1]
```diff
@@ -0,0 +1,12 @@
+set(UHDM_VERSION "@PROJECT_VERSION_MAJOR@.@PROJECT_VERSION_MINOR@")
+
+@PACKAGE_INIT@
+
+include(${CMAKE_CURRENT_LIST_DIR}/SurelogTargets.cmake)
+
+set_and_check(SURELOG_BIN_DIR @CMAKE_INSTALL_PREFIX@/bin)
+set_and_check(SURELOG_INCLUDE_DIR @CMAKE_INSTALL_PREFIX@/include/Surelog)
+set_and_check(SURELOG_INCLUDE_DIRS @CMAKE_INSTALL_PREFIX@/include/Surelog)
+set_and_check(SURELOG_LIB_DIR @CMAKE_INSTALL_PREFIX@/lib/surelog)
```
Comment (hzeller) [inline cmake/configs/SurelogConfig.cmake.in:10]: Do do the proper lib64 situation that we also had to fix in pkg-config, wouldn't this be `@CMAKE_INSTALL_FULL_LIBDIR@/surelog` ?

**Comment by hzeller** [comment]
To test it locally, I had to patch `SurelogConfig.cmake.in` to use the cmake install library path.
And then, to test with the makefile `test_install`, I added the prefix to that command line as well:

```patch
--- a/Makefile
+++ b/Makefile
@@ -134,7 +134,7 @@ install-shared: release-shared
        cmake --install build
 
 test_install:
-       cmake -DCMAKE_BUILD_TYPE=Release -DINSTALL_DIR=$(PREFIX) -S tests/TestInstall -B tests/TestInstall/build
+       cmake -DCMAKE_BUILD_TYPE=Release -DINSTALL_DIR=$(PREFIX) -DCMAKE_PREFIX_PATH=$(PREFIX) $(ADDITIONAL_CMAKE_OPTIONS) -S tests/TestInstall -B tests/TestInstall/build
        cmake --build tests/TestInstall/build -j $(CPU_CORES)
 
 # Using pkg-config. Its search-path might be set in different ways. Set both.
diff --git a/cmake/configs/SurelogConfig.cmake.in b/cmake/configs/SurelogConfig.cmake.in
index 494db77cb..c4d7e7ad1 100644
--- a/cmake/configs/SurelogConfig.cmake.in
+++ b/cmake/configs/SurelogConfig.cmake.in
@@ -7,6 +7,6 @@ include(${CMAKE_CURRENT_LIST_DIR}/SurelogTargets.cmake)
 set_and_check(SURELOG_BIN_DIR @CMAKE_INSTALL_PREFIX@/bin)
 set_and_check(SURELOG_INCLUDE_DIR @CMAKE_INSTALL_PREFIX@/include/Surelog)
 set_and_check(SURELOG_INCLUDE_DIRS @CMAKE_INSTALL_PREFIX@/include/Surelog)
-set_and_check(SURELOG_LIB_DIR @CMAKE_INSTALL_PREFIX@/lib/surelog)
+set_and_check(SURELOG_LIB_DIR @CMAKE_INSTALL_FULL_LIBDIR@/surelog)
 
 check_required_components(Surelog)
```

Also, I changed the ` @CMAKE_INSTALL_FULL_LIBDIR@` in the UHDM config as well:
https://github.com/chipsalliance/UHDM/blob/master/cmake/configs/UHDMConfig.cmake.in#L10

I am using host gtest and flatbuffers, and the UHDM from third_party/
ADDITIONAL_CMAKE_OPTIONS="-DSURELOG_USE_HOST_GTEST=On -DSURELOG_USE_HOST_FLATBUFFERS=On -DSURELOG_USE_HOST_ANTLR=On"

Running a test install from the makefile, I now see that it does not find the UHDM library:
```
PREFIX=/tmp/test-3557-1 make install
# ... works fine
PREFIX=/tmp/test-3557-1 make test_install
cmake -DCMAKE_BUILD_TYPE=Release -DINSTALL_DIR=/tmp/test-3557-1 -DCMAKE_PREFIX_PATH=/tmp/test-3557-1 -DSURELOG_USE_HOST_GTEST=On -DSURELOG_USE_HOST_FLATBUFFERS=On -DSURELOG_USE_HOST_ANTLR=On -DANTLR_JAR_LOCATION=/nix/store/hlzfp8rynjw6gdhzdlzwy6dj06qw8s6p-antlr-4.12.0/share/java/antlr-4.12.0-complete.jar -S tests/TestInstall -B tests/TestInstall/build
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /nix/store/h5003wsy3qqimqvrkn3bc5mwq4hhidag-gcc-wrapper-12.2.0/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /nix/store/h5003wsy3qqimqvrkn3bc5mwq4hhidag-gcc-wrapper-12.2.0/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE  
-- Found ANTLR: /nix/store/hlzfp8rynjw6gdhzdlzwy6dj06qw8s6p-antlr-4.12.0/share/java/antlr-4.12.0-complete.jar (found version "4.12.0") 
-- Found GTest: /nix/store/ak9zm2b7k61di6fnrmw0kcndh7hi75k0-gtest-1.12.1-dev/lib/cmake/GTest/GTestConfig.cmake (found version "1.12.1")  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/hzeller/src/my/Surelog/tests/TestInstall/build
cmake --build tests/TestInstall/build -j 20
make[1]: Entering directory '/home/hzeller/src/my/Surelog/tests/TestInstall/build'
make[1]: warning: -j20 forced in submake: resetting jobserver mode.
[ 50%] Building CXX object CMakeFiles/test_hellosureworld.dir/home/hzeller/src/my/Surelog/src/hellosureworld.cpp.o
[100%] Linking CXX executable test_hellosureworld
/nix/store/8qm6sjqa09a03glzmafprpp69k74l4lm-binutils-2.40/bin/ld: cannot find -luhdm: No such file or directory
collect2: error: ld returned 1 exit status
make[3]: *** [CMakeFiles/test_hellosureworld.dir/build.make:104: test_hellosureworld] Error 1
make[2]: *** [CMakeFiles/Makefile2:85: CMakeFiles/test_hellosureworld.dir/all] Error 2
make[1]: *** [Makefile:91: all] Error 2
make[1]: Leaving directory '/home/hzeller/src/my/Surelog/tests/TestInstall/build'
make: *** [Makefile:138: test_install] Error 2
```

The install does include the UHDM *.cmake files:
```
$ find /tmp/test-3557/ -name "*.cmake"
/tmp/test-3557/lib64/cmake/UHDM/UHDMTargets-release.cmake
/tmp/test-3557/lib64/cmake/UHDM/UHDMConfigVersion.cmake
/tmp/test-3557/lib64/cmake/UHDM/UHDMTargets.cmake
/tmp/test-3557/lib64/cmake/UHDM/UHDMConfig.cmake
/tmp/test-3557/lib64/cmake/Surelog/SurelogTargets.cmake
/tmp/test-3557/lib64/cmake/Surelog/SurelogTargets-release.cmake
/tmp/test-3557/lib64/cmake/Surelog/SurelogConfig.cmake
/tmp/test-3557/lib64/cmake/Surelog/SurelogConfigVersion.cmake
```

So I suspect some find_package() or something is missing in this case ?

File: cmake/configs/SurelogConfig.cmake.in [hunk 2]
```diff
@@ -0,0 +1,12 @@
+set(UHDM_VERSION "@PROJECT_VERSION_MAJOR@.@PROJECT_VERSION_MINOR@")
```
Comment (hzeller) [inline cmake/configs/SurelogConfig.cmake.in:1]: `UHDM_VERSION` ?

**Comment by hzeller** [comment]
the library path change for UHDM is in https://github.com/chipsalliance/UHDM/pull/915

**Comment by hzeller** [comment]
The CI issue on MSYS can probably be fixed with the Makefile change [from above](https://github.com/chipsalliance/Surelog/pull/3557#issuecomment-1507666837).

The other Windows issue ... don't know that looks like some issue due to backslashes in Windows paths (I wonder how many trillion lost GDP the choice to use an escape character as path separator from 40 years ago has cost society ...).

File: tests/TestInstall/CMakeLists.txt [hunk 3]
```diff
@@ -40,9 +49,11 @@ endif()
 if(SURELOG_USE_HOST_UHDM)
   find_package(UHDM REQUIRED)
   find_package(CapnProto)
+  set(UHDM_LIBRARY uhdm::uhdm)
   set(CAPNPROTO_LIBRARY CapnProto::capnp)
 else()
   set(UHDM_LIBRARY uhdm)
+  set(UHDM_LIB_DIR "${INSTALL_DIR}/lib/uhdm")
```
Comment (hzeller) [inline tests/TestInstall/CMakeLists.txt:56]: On my system, the library ends up in lib64/, so a local hack of
```
set(UHDM_LIB_DIR "${INSTALL_DIR}/lib/uhdm" "${INSTALL_DIR}/lib64/uhdm")
```

does work. But maybe there are better ways ?

Comment (timkpaine) [reply tests/TestInstall/CMakeLists.txt:56]: it ends up there when installed alongside surelog (not separately)?

Comment (timkpaine) [reply tests/TestInstall/CMakeLists.txt:56]: im cool with that hack, its possible that `find_package(UHDM)` may still work/cmake file still installed when its vendored, I can double check the logs

Comment (hzeller) [reply tests/TestInstall/CMakeLists.txt:56]: Yes, the only thing that ends up in lib is UHDM.capnp. Might be a particular configuration of cmake on that 64 bit system.

Alright, so let's put in this hack and see how things go.

**Comment by hzeller** [comment]
FYI, updated UHDM in third_party/ in #3589

**Review by hzeller — APPROVED** [review]
Nice!
Thanks for pushing through with this!

**Comment by timkpaine** [comment]
thanks! what a relief, that was a pain...but hopefully its easier to build on top of/integrate with now

**Comment by hzeller** [comment]
Tagged 1.55 in #3590

**Pull request merged**